bgpd/bgp_label.c

```diff
@@ -198,32 +198,40 @@ int bgp_parse_fec_update(void)
 	return 1;
 }
 
-mpls_label_t bgp_adv_label(struct bgp_dest *dest, struct bgp_path_info *pi,
-			   struct peer *to, afi_t afi, safi_t safi)
+void bgp_adv_label(struct bgp_dest *dest, struct bgp_path_info *pi, struct peer *to, afi_t afi,
+		   safi_t safi, mpls_label_t *labels, uint8_t *num_labels)
 {
 	struct peer *from;
-	mpls_label_t remote_label;
 	int reflect;
+	uint8_t i;
+	bool use_local_label = true;
 
-	if (!dest || !pi || !to)
-		return MPLS_INVALID_LABEL;
+	labels[0] = MPLS_INVALID_LABEL;
+	if (!dest || !pi || !to) {
+		*num_labels = 0;
+		return;
+	}
 
-	remote_label = BGP_PATH_INFO_NUM_LABELS(pi)
-			       ? pi->extra->labels->label[0]
-			       : MPLS_INVALID_LABEL;
 	from = pi->peer;
 	reflect =
 		((from->sort == BGP_PEER_IBGP) && (to->sort == BGP_PEER_IBGP));
 
 	if (reflect
 	    && !CHECK_FLAG(to->af_flags[afi][safi],
 			   PEER_FLAG_FORCE_NEXTHOP_SELF))
-		return remote_label;
+		use_local_label = false;
 
 	if (CHECK_FLAG(to->af_flags[afi][safi], PEER_FLAG_NEXTHOP_UNCHANGED))
-		return remote_label;
+		use_local_label = false;
 
-	return dest->local_label;
+	if (use_local_label) {
+		*num_labels = 1;
+		labels[0] = dest->local_label;
+	} else {
```

Comment:
is there a BOS issue anywhere in here? do we expect that BOS is already set for outgoing labels?